Package/ICustomPackageAction.cs

```diff
@@ -101,22 +101,38 @@ internal static void RunCustomActions(PackageDef package, PackageActionStage sta
 
             log.Debug($"Available custom actions for '{stage.ToString().ToLower()}' stage. ({customActions.Count} actions: {string.Join(", ", customActions.Select(s => s.ToString()))})");
 
-            foreach (ICustomPackageAction action in customActions)
+            try
             {
-                Stopwatch timer = Stopwatch.StartNew();
-                try
+                // Allow child processes to bypass the lock on the installation which is held by this process.
+                // We need to set this on the User level instead of the Process level because there are plugins predating
+                // the locking feature that depend on this behavior, such as 'OSIntegration'.
+                // OSIntegration starts a 'tap.exe' subprocess with the 'runas' verb on order to run as administrator.
+                // When a process is started with the 'runas' verb, the initiating process' environment is not inherited.
+                Environment.SetEnvironmentVariable(FileLock.InstallationLockEnv, "1", EnvironmentVariableTarget.User);
+                // Also set it on the process to avoid potential edgecases with other platforms that don't have three different environments.
+                Environment.SetEnvironmentVariable(FileLock.InstallationLockEnv, "1", EnvironmentVariableTarget.Process);
+                foreach (ICustomPackageAction action in customActions)
                 {
-                    if (action.Execute(package, args))
+                    Stopwatch timer = Stopwatch.StartNew();
+                    try
                     {
-                        log.Info(timer, $"Package action {action.GetType().Name} completed");
-                        continue;
+                        if (action.Execute(package, args))
+                        {
+                            log.Info(timer, $"Package action {action.GetType().Name} completed");
+                            continue;
+                        }
+                    }
+                    catch (Exception ex)
+                    {
+                        log.Warning(timer, $"Package action {action.ToString()} failed", ex);
+                        throw;
                     }
                 }
-                catch (Exception ex)
-                {
-                    log.Warning(timer, $"Package action {action.ToString()} failed", ex);
-                    throw;
-                }
+            }
+            finally
+            {
+                Environment.SetEnvironmentVariable(FileLock.InstallationLockEnv, null, EnvironmentVariableTarget.User);
```

Comment:
Here you are assuming that the environment variable was null before it was set to "1". Should this maybe be reverted instead to the previous value to avoid unintended side effects?

Reply:
I thought about it but I didn't want to risk a race condition where it will never be set back to null.